StartBonusActivation/handler.ts

```diff
@@ -369,68 +369,59 @@ export function StartBonusActivationHandler(
-        }))
+      .chainSecond(checkEligibilityCheckIsRunning(dfClient, fiscalCode))
+      .chainSecond(checkBonusActivationIsRunning(dfClient, fiscalCode))
+      .chainSecond(
+        getLatestValidDSU(eligibilityCheckModel, fiscalCode).map(
+          (dsu: Dsu) => ({
+            dsu,
+            familyUID: generateFamilyUID(dsu.familyMembers)
+          })
+        )
       )
-      .chain<BonusActivationWithFamilyUID>(({ dsu, familyUID }) => {
-        // this part is on a sub-chain as it handles the lock/unlock protection mechanism
-        // familyUID serves as lock context and thus is needed in scope for every sub-task
-        return (
-          taskEither
-            .of<StartBonusActivationResponse, void>(void 0)
-
-            .chain(_ =>
-              createBonusActivation(
-                bonusActivationModel,
-                fiscalCode,
-                familyUID,
-                dsu
-              )
-            )
-            .chain(bonusActivation =>
+      .chain<BonusActivationWithFamilyUID>(({ dsu, familyUID }) =>
+        bracket(
+          acquireLockForUserFamily(bonusLeaseModel, familyUID).mapLeft(l => {
+            defaultClient.trackException({
+              exception: new Error(l.detail),
+              properties: {
+                name: "bonus.activation.lock.acquire"
+              }
+            });
+            return l;
+          }),
+          _ => {
+            // this part is on a sub-chain as it handles the lock/unlock protection mechanism
+            // familyUID serves as lock context and thus is needed in scope for every sub-task
+            return createBonusActivation(
+              bonusActivationModel,
+              fiscalCode,
+              familyUID,
+              dsu
+            ).chain(bonusActivation =>
               runStartBonusActivationOrchestrator(
-                client,
+                dfClient,
                 bonusActivation,
                 fiscalCode
-              ).map(_ => bonusActivation)
-            )
-            // the following is basically:
-            // on right, just pass it
-            // on left, perform unlock but then pass the original left value
-            .foldTaskEither(
-              l => {
-                defaultClient.trackException({
-                  exception: new Error(l.detail),
-                  properties: {
-                    name: "bonus.activation.start"
-                  }
-                });
-                return relaseLockForUserFamily(
-                  bonusLeaseModel,
-                  familyUID
-                ).foldTaskEither(
-                  _ => fromEither(left(l)),
-                  _ => fromEither(left(l))
-                );
-              },
-              r => fromEither(right(r))
-            )
-        );
-      })
+              ).map(__ => bonusActivation)
+            );
+          },
+          _ =>
+            relaseLockForUserFamily(bonusLeaseModel, familyUID).mapLeft(l => {
```

Comment:
we don't want to relase the lock in case the body (createBonusActivation + runStartBonusActivationOrchestrator) success 

if we do, users could request a bonus for the same family(UID)

that's the reason `relaseLockForUserFamily` is called _only_ when one between `createBonusActivation` or `runStartBonusActivationOrchestrator` fail in the original code

Reply:
I see, the naming here is a little bit misleading, I thought the lease/lock was an actual lease (that would eventually expire) similar to the blob lease but is indeed an actual document that states that a bonus exist for that familyUID

Reply:
yes that's because the original implementation differs from the current one (it was a temporary lock). we diverged because the previous implementation rely on the fact that you can have a unique key (familyUID) on the Bonus stored into cosmosdb, but we discovered that is not possible to have unique keys with cosmosdb (apart from the 'id' field which is the secret bonus code)